### ISSUE TITLE
Effect of changepoint_prior_scale and seasonality_prior_scale

### ISSUE THREAD

**Issue by DivJ** [facebook/prophet#1320]
Hi,
I am using prophet in one of my problems.
One thing I have noticed is the effect of changepoint_prior_scale and seasonality_prior_scale on the output. At present I am using hyperopt and cross-validation mape to find optimum parameters. Below is the grid I am using:
`space = {   'changepoint_prior_scale': hp.uniform('changepoint_prior_scale', 0.001, 1),
     'seasonality_prior_scale': hp.uniform('seasonality_prior_scale', 0.001, 1),
     'seasonality_mode': hp.choice('seasonality_mode', ['additive','multiplicative'])
    } 
tpe_best = fmin(fn=objective_prophet_min,     
  space=space, 
                algo=tpe.suggest, trials=Trials(), 
                max_evals=15)`
My forecast changed from 63 to 178 while running the same iteration 2 times (Actual value is 180).
I understand the impact of choosing the correct parameters but such a huge amount of change is problematic as we cannot trust whether the next time the model produces forecast, will they be reliable or just a matter of chance.
Any ideas/suggestions on that?

**Comment by bletham** [comment]
Short time series can definitely have sensitivity to the parameters, particularly the changepoint_prior_scale. This can significantly impact the prediction: A very small value will force a linear trend, while a large value will allow the trend to fluctuate quite a bit (and, if there is little data, possibly capture seasonal effects). Cross validation is the right way to select which ones to use. For the changepoint_prior_scale, I think the uncertainty intervals should give you a pretty good idea of how reliable it is. If changepoint_prior_scale has a very high value, then it will allow for a lot of trend flexibility, it will fit the history better, but there will be more uncertainty in the future as reflected by wider uncertainty intervals.

A few additional comments:
- Hyperopt with tpe does model-based optimization. max_evals=15 is pretty small for a 3-dim space, though it's probably slow because of model fitting time. With a reasonable number of cutoffs, the cross validation will be pretty fast and I think you'd be better off doing a larger number (50?) of random search iterations than TPE.
- I think it would probably make the most sense to use a log scaling for the prior scales rather than linear (uniform).
- The seasonality mode is usually pretty clear from the data. Data where the seasonality affects a % of users (probably most cases) will be multiplicative. It's pretty easy to tell from looking at a time series if it has multiplicative or additive seasonality (see e.g. https://facebook.github.io/prophet/docs/multiplicative_seasonality.html). Generally, I would highly recommend plotting the time series if you are concerned with validity of the hyperparameters. A similar thing can be said for the changepoint prior scale: If the historical trend is being over or under fit, it's usually pretty easy to tell by looking at it.

**Comment by DivJ** [comment]
Thank you @bletham for the suggestions. I have modified my grid as below:

```
 space =  {
     'changepoint_prior_scale': hp.uniform('changepoint_prior_scale', -3,1),
     'seasonality_prior_scale': hp.uniform('seasonality_prior_scale', -3,1),
     'seasonality_mode': hp.choice('seasonality_mode', ['additive','multiplicative'])
    }

objective_prophet_min=partial(objective_prophet,nodeForecast=nodeForecast,rescale_factor=rescale_factor)
    tpe_best = fmin(fn=objective_prophet_min, space=space, 
                algo=rand.suggest, trials=Trials(), 
                max_evals=25)

```

Since I am forecasting around 600 series, I have added seasonality as a parameter and I cannot visualize each and every time series to check the parameter effect.

Below is how I am doing cross-validation: (I have only 2 years worth of data)
```
  result=cross_validation(model, initial=pd.to_timedelta(365,unit="d"), 
                            period=pd.to_timedelta(180,unit="d"), 
                            horizon = pd.to_timedelta(180,unit="d"))
```
Also I am using box-cox transformations on my additional regressors as well as target.
Do you have any other advice which can help me to improve the results?

BTW, amazing work on prophet. Just loved this model.

**Comment by bletham** [comment]
That all seems good to me! I don't have any other general advice, just if there are particular forecasts that seem degenerate to you I could comment on what may be happening.

**Comment by DivJ** [comment]
HI @bletham,
I am running the above code but still my forecast swing is too high. It swings from 170 to 400 with different params.
Also do youhave any tips for regressor selection?

**Comment by bletham** [comment]
can you post the `m.plot` plot for one of these forecasts with a large swing, under each of the extreme sets of parameters?

**Comment by DivJ** [comment]
Hi @bletham 

**Actuals : 179**
**Prediction : 143 {'changepoint_prior_scale': -1.5226686219548333, 'seasonality_prior_scale': -1.048897012671966}**
![143](https://user-images.githubusercontent.com/19840519/76215672-41814100-6235-11ea-8ba4-ec97961e5739.png)

**Prediction: 436 {'changepoint_prior_scale': -1.5672635400686978, 'seasonality_prior_scale': -1.8225518036711934}**
![436](https://user-images.githubusercontent.com/19840519/76215782-71c8df80-6235-11ea-901e-a929cba01e56.png)

I found these paramters using cross-validation code above. SInce I am doing forecasts for a lot of series, Can you suggest a way for me to automate this? I can see that the uncertainity in trend is really high in second case.

**Comment by bletham** [comment]
Could you post the components plot also, and include the points in the history too? I'm trying to figure out how seasonality/trend are being balanced out in fit.
I think part of what might be happening here is that with monthly data and only 2 years it isn't a ton of data points relative to the model parameters, which makes the model more sensitive to the fitting / hyperparameters than I'd expect it to be with more data points. That said, there is a really clear yearly seasonality here.

One thing you could try to do to reduce the number of parameters in the model (and thus likely also reduce sensitivity to hyperparameters) is to replace the built-in yearly seasonality with a binary regressor for each month. This would make sense specifically because it is monthly data here (related to #1168). Specifically, you'd add an extra regressor `is_jan` that is 1 for the Jan datapoints and 0 otherwise. And similarly for the other 11 months, for 12 total extra regressors. And then disable the built-in yearly seasonality. That might make things more stable, and make it easier to use the cross validation approach for automatic parameter selection.

Generally I think cross validation makes a lot of sense for parameter selection, I think the only other thing you might consider to improve robustness would be to find one set of hyperparameters for all 600 time series, by optimizing for the aggregated error across all time series. This would make sense if these time series are all similar in nature, and if you were able to aggregate the errors in a good way. You can think about this as applying regularization to the hyperparameters to reduce overfitting them on some time series (at the cost of potentially underfitting on others).

**Comment by DivJ** [comment]
Thanks for quick reply. I will definitely try that and post the results.
I was also thinking of checking the band gap between yhat-lower and yhat-upper and if the difference is above a particular threshold percentage then reject that and rerun the cross valid again. Does that make any sense?
